File: ConcurrentCppReadersWriters/src/ReadersWriters.cpp
/*
 * ReadersWriters.cpp
 *
 *  Created on: May 9, 2010
 *      Author: Rafael de Castro Dantas Sales - rafaelcds@gmail.com
 */

#include <iostream>
#include <pthread.h>
#include "ReadersWriters.h"

using namespace std;

ReadersWriters::ReadersWriters() {
	data = 0;
	numReaders = 0;
}

ReadersWriters::~ReadersWriters() {
}

ReadersWriters::Reader::Reader(ReadersWriters& readersWriters, int readerId) :
	readersWriters(readersWriters), readerId(readerId) {
}

void ReadersWriters::Reader::read() {
	startRead();
	cout << "Reader No." << readerId << ": read " << readersWriters.data << endl;
	endRead();
}

void ReadersWriters::Reader::read(int numRounds) {
	for (int i = 0; i < numRounds; i++) {
		read();
	}
}

void ReadersWriters::Reader::startRead() {
	pthread_mutex_lock(&readersWriters.lock);
	readersWriters.numReaders++;
	pthread_mutex_unlock(&readersWriters.lock);
}

void ReadersWriters::Reader::endRead() {
	pthread_mutex_lock(&readersWriters.lock);
	readersWriters.numReaders--;
	if (readersWriters.numReaders == 0) {
		pthread_cond_signal(&readersWriters.canWrite);
	}
	pthread_mutex_unlock(&readersWriters.lock);
}

ReadersWriters::Writer::Writer(ReadersWriters& readersWriters, int writerId) :
	readersWriters(readersWriters), writerId(writerId) {
}

void ReadersWriters::Writer::write() {
	pthread_mutex_lock(&readersWriters.lock);
	while (readersWriters.numReaders > 0) {
		pthread_cond_wait(&readersWriters.canWrite, &readersWriters.lock);
	}
	readersWriters.data++;
	cout << "Writer No." << writerId << ": wrote " << readersWriters.data << endl;
	pthread_cond_signal(&readersWriters.canWrite);
	pthread_mutex_unlock(&readersWriters.lock);
}

void ReadersWriters::Writer::write(int numRounds) {
	for (int i = 0; i < numRounds; i++) {
		write();
	}
}

File: ConcurrentJavaReadersWriters/src/rafael/concurrent/monitor/readerswriters/ReadersWriters.java
package rafael.concurrent.monitor.readerswriters;

/**
 * This class is an implementation of Readers and Writers concurrency problem
 * 
 * @author Rafael de Castro Dantas Sales - rafaelcds@gmail.com
 */
public class ReadersWriters {

	private int data = 0;
	private int numReaders = 0;

	public class Reader {

		private int readerId;

		public Reader(int readerId) {
			this.readerId = readerId;
		}

		private void startRead() {
			synchronized (ReadersWriters.this) {
				numReaders++;
			}
		}

		private void endRead() {
			synchronized (ReadersWriters.this) {
				numReaders--;
				if (numReaders == 0) {
					ReadersWriters.this.notify();
				}
			}
		}

		public void read() {
			startRead();
			System.out.println("Reader No." + readerId + ": read " + data);
			endRead();
		}

		public void read(int numRounds) {
			for (int i = 0; i < numRounds; i++) {
				read();
			}
		}
	}

	public class Writer {

		private int writerId;

		public Writer(int writerId) {
			this.writerId = writerId;
		}

		public void write() {
			synchronized (ReadersWriters.this) {
				while (numReaders > 0) {
					try {
						ReadersWriters.this.wait();
					} catch (InterruptedException ex) {
						return;
					}
				}
				data++;
				System.out.println("Writer No." + writerId + ": wrote " + data);
				ReadersWriters.this.notify();
			}
		}

		public void write(int numRounds) {
			for (int i = 0; i < numRounds; i++) {
				write();
			}
		}
	}
}
